File: airbyte-integrations/connectors/destination-teradata/src/main/java/io/airbyte/integrations/destination/teradata/TeradataDestination.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.integrations.destination.teradata;

import com.fasterxml.jackson.databind.JsonNode;
import com.google.common.collect.ImmutableMap;
import io.airbyte.cdk.db.factory.DatabaseDriver;
import io.airbyte.cdk.db.jdbc.JdbcDatabase;
import io.airbyte.cdk.db.jdbc.JdbcUtils;
import io.airbyte.cdk.integrations.base.Destination;
import io.airbyte.cdk.integrations.base.IntegrationRunner;
import io.airbyte.cdk.integrations.destination.NamingConventionTransformer;
import io.airbyte.cdk.integrations.destination.StandardNameTransformer;
import io.airbyte.cdk.integrations.destination.async.deser.StreamAwareDataTransformer;
import io.airbyte.cdk.integrations.destination.jdbc.AbstractJdbcDestination;
import io.airbyte.cdk.integrations.destination.jdbc.typing_deduping.JdbcDestinationHandler;
import io.airbyte.cdk.integrations.destination.jdbc.typing_deduping.JdbcSqlGenerator;
import io.airbyte.commons.json.Jsons;
import io.airbyte.integrations.base.destination.typing_deduping.DestinationHandler;
import io.airbyte.integrations.base.destination.typing_deduping.ParsedCatalog;
import io.airbyte.integrations.base.destination.typing_deduping.SqlGenerator;
import io.airbyte.integrations.base.destination.typing_deduping.migrators.Migration;
import io.airbyte.integrations.base.destination.typing_deduping.migrators.MinimumDestinationState;
import io.airbyte.integrations.destination.teradata.typing_deduping.TeradataDataTransformer;
import io.airbyte.integrations.destination.teradata.typing_deduping.TeradataDestinationHandler;
import io.airbyte.integrations.destination.teradata.typing_deduping.TeradataRawTableAirbyteMetaMigration;
import io.airbyte.integrations.destination.teradata.typing_deduping.TeradataSqlGenerator;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.io.PrintWriter;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

/**
  Main class for hte destination-teradata connector
 */
public class TeradataDestination extends AbstractJdbcDestination<MinimumDestinationState> implements Destination {

  private static final Logger LOGGER = LoggerFactory.getLogger(TeradataDestination.class);

  private static final NamingConventionTransformer NAMING_CONVENTION_TRANSFORMER = new StandardNameTransformer();

  /**
   * Teradata JDBC driver
   */
  public static final String DRIVER_CLASS = DatabaseDriver.TERADATA.getDriverClassName();
  /**
   * Default schema name
   */
  protected static final String DEFAULT_SCHEMA_NAME = "def_airbyte_db";
  protected static final String PARAM_MODE = "mode";
  protected static final String PARAM_SSL = "ssl";
  protected static final String PARAM_SSL_MODE = "ssl_mode";
  protected static final String PARAM_SSLMODE = "sslmode";
  protected static final String PARAM_SSLCA = "sslca";
  protected static final String REQUIRE = "require";

  protected static final String VERIFY_CA = "verify-ca";

  protected static final String VERIFY_FULL = "verify-full";

  protected static final String ALLOW = "allow";

  protected static final String CA_CERTIFICATE = "ca.pem";

  protected static final String CA_CERT_KEY = "ssl_ca_certificate";

  protected static final String ENCRYPTDATA = "ENCRYPTDATA";

  protected static final String ENCRYPTDATA_ON = "ON";
  private static final String DROP_CASCADE_OPTION = "drop_cascade";


  public static void main(String[] args) throws Exception {
    new IntegrationRunner(new TeradataDestination()).run(args);
  }

  public TeradataDestination() {
    super(DRIVER_CLASS, new StandardNameTransformer(), new TeradataSqlOperations());
  }

  private static void createCertificateFile(String fileName, String fileValue) throws IOException {
    try (final PrintWriter out = new PrintWriter(fileName, StandardCharsets.UTF_8)) {
      out.print(fileValue);
    }
  }

  @Override
  protected Map<String, String> getDefaultConnectionProperties(final JsonNode config) {
    final Map<String, String> additionalParameters = new HashMap<>();
    if (config.has(PARAM_SSL) && config.get(PARAM_SSL).asBoolean()) {
      LOGGER.debug("SSL Enabled");
      if (config.has(PARAM_SSL_MODE)) {
        LOGGER.debug("Selected SSL Mode : " + config.get(PARAM_SSL_MODE).get(PARAM_MODE).asText());
        additionalParameters.putAll(obtainConnectionOptions(config.get(PARAM_SSL_MODE)));
      } else {
        additionalParameters.put(PARAM_SSLMODE, REQUIRE);
      }
    }
    additionalParameters.put(ENCRYPTDATA, ENCRYPTDATA_ON);
    return additionalParameters;
  }

  private Map<String, String> obtainConnectionOptions(final JsonNode encryption) {
    final Map<String, String> additionalParameters = new HashMap<>();
    if (!encryption.isNull()) {
      final var method = encryption.get(PARAM_MODE).asText();
      switch (method) {
        case "verify-ca", "verify-full" -> {
          additionalParameters.put(PARAM_SSLMODE, method);
          try {
            createCertificateFile(CA_CERTIFICATE, encryption.get("ssl_ca_certificate").asText());
          } catch (final IOException ioe) {
            throw new RuntimeException("Failed to create certificate file");
          }
          additionalParameters.put(PARAM_SSLCA, CA_CERTIFICATE);
        }
        default -> {
          additionalParameters.put(PARAM_SSLMODE, method);
        }
      }
    }
    return additionalParameters;
  }

  @Override
  public JsonNode toJdbcConfig(final JsonNode config) {
    final String schema = Optional.ofNullable(config.get(JdbcUtils.SCHEMA_KEY)).map(JsonNode::asText).orElse(DEFAULT_SCHEMA_NAME);

    final String jdbcUrl = String.format("jdbc:teradata://%s/",
        config.get(JdbcUtils.HOST_KEY).asText());

    final ImmutableMap.Builder<Object, Object> configBuilder = ImmutableMap.builder()
        .put(JdbcUtils.USERNAME_KEY, config.get(JdbcUtils.USERNAME_KEY).asText())
        .put(JdbcUtils.JDBC_URL_KEY, jdbcUrl)
        .put(JdbcUtils.SCHEMA_KEY, schema);

    if (config.has(JdbcUtils.PASSWORD_KEY)) {
      configBuilder.put(JdbcUtils.PASSWORD_KEY, config.get(JdbcUtils.PASSWORD_KEY).asText());
    }

    if (config.has(JdbcUtils.JDBC_URL_PARAMS_KEY)) {
      configBuilder.put(JdbcUtils.JDBC_URL_PARAMS_KEY, config.get(JdbcUtils.JDBC_URL_PARAMS_KEY).asText());
    }
    return Jsons.jsonNode(configBuilder.build());
  }

  @Override
  public boolean isV2Destination() {
    return true;
  }


  @Override
  protected JdbcSqlGenerator getSqlGenerator(final JsonNode config) {
    final JsonNode dropCascadeNode = config.get(DROP_CASCADE_OPTION);
    final boolean dropCascade = dropCascadeNode != null && dropCascadeNode.asBoolean();
    return new TeradataSqlGenerator(new TeradataNameTransformer(), dropCascade);
  }

  @Override
  protected JdbcDestinationHandler<MinimumDestinationState> getDestinationHandler(String databaseName, JdbcDatabase database, String rawTableSchema) {
    return new TeradataDestinationHandler(databaseName, database, rawTableSchema);
  }

  @Override
  protected List<Migration<MinimumDestinationState>> getMigrations(JdbcDatabase database,
                                                         String databaseName,
                                                         SqlGenerator sqlGenerator,
                                                         DestinationHandler<MinimumDestinationState> destinationHandler) {
    return List.of(new TeradataRawTableAirbyteMetaMigration(database, databaseName));
  }

  @Override
  protected StreamAwareDataTransformer getDataTransformer(ParsedCatalog parsedCatalog, String defaultNamespace) {
    return new TeradataDataTransformer();
  }
}

File: airbyte-integrations/connectors/destination-teradata/src/main/kotlin/io/airbyte/integrations/destination/teradata/TeradataDestination.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.integrations.destination.teradata

import com.fasterxml.jackson.databind.JsonNode
import com.google.common.collect.ImmutableMap
import io.airbyte.cdk.db.factory.DataSourceFactory
import io.airbyte.cdk.db.jdbc.DefaultJdbcDatabase
import io.airbyte.cdk.db.jdbc.JdbcDatabase
import io.airbyte.cdk.db.jdbc.JdbcUtils
import io.airbyte.cdk.integrations.base.Destination
import io.airbyte.cdk.integrations.base.IntegrationRunner
import io.airbyte.cdk.integrations.destination.StandardNameTransformer
import io.airbyte.cdk.integrations.destination.jdbc.AbstractJdbcDestination
import io.airbyte.cdk.integrations.destination.jdbc.JdbcGenerationHandler
import io.airbyte.cdk.integrations.destination.jdbc.SqlOperations
import io.airbyte.cdk.integrations.destination.jdbc.typing_deduping.JdbcDestinationHandler
import io.airbyte.cdk.integrations.destination.jdbc.typing_deduping.JdbcSqlGenerator
import io.airbyte.commons.json.Jsons
import io.airbyte.commons.map.MoreMaps
import io.airbyte.integrations.base.destination.typing_deduping.DestinationHandler
import io.airbyte.integrations.base.destination.typing_deduping.DestinationV1V2Migrator
import io.airbyte.integrations.base.destination.typing_deduping.SqlGenerator
import io.airbyte.integrations.base.destination.typing_deduping.migrators.Migration
import io.airbyte.integrations.base.destination.typing_deduping.migrators.MinimumDestinationState
import io.airbyte.integrations.destination.teradata.typing_deduping.TeradataDestinationHandler
import io.airbyte.integrations.destination.teradata.typing_deduping.TeradataGenerationHandler
import io.airbyte.integrations.destination.teradata.typing_deduping.TeradataSqlGenerator
import io.airbyte.integrations.destination.teradata.typing_deduping.TeradataV1V2Migrator
import io.airbyte.integrations.destination.teradata.util.TeradataConstants
import java.io.IOException
import java.io.PrintWriter
import java.nio.charset.StandardCharsets
import java.util.*
import java.util.regex.Pattern
import javax.sql.DataSource
import kotlin.collections.HashMap

/**
 * The TeradataDestination class is responsible for handling the connection to the Teradata database
 * as destination from Airbyte. It extends the AbstractJdbcDestination class and implements the
 * Destination interface, facilitating the configuration and management of database interactions,
 * including setting a query band.
 */
class TeradataDestination :
    AbstractJdbcDestination<MinimumDestinationState>(
        TeradataConstants.DRIVER_CLASS,
        StandardNameTransformer()
    ),
    Destination {
    /**
     * Retrieves the data source for the Teradata database connection.
     *
     * @param config The configuration settings as a JsonNode.
     * @return The DataSource object for the Teradata connection.
     */
    override fun getDataSource(config: JsonNode): DataSource {
        val jdbcConfig = toJdbcConfig(config)
        val connectionProperties = getConnectionProperties(config)
        val dataSource =
            DataSourceFactory.create(
                jdbcConfig[JdbcUtils.USERNAME_KEY]?.asText(),
                jdbcConfig[JdbcUtils.PASSWORD_KEY]?.asText(),
                TeradataConstants.DRIVER_CLASS,
                jdbcConfig[JdbcUtils.JDBC_URL_KEY].asText(),
                connectionProperties,
                getConnectionTimeout(connectionProperties)
            )
        // set session query band
        setQueryBand(getDatabase(dataSource))
        return dataSource
    }

    public override fun getConnectionProperties(config: JsonNode): Map<String, String> {
        return MoreMaps.merge(appendLogMech(config), super.getConnectionProperties(config))
    }
    /** Appends Logging Mechanism to JDBC URL */
    private fun appendLogMech(config: JsonNode): Map<String, String> {
        val logmechParams: MutableMap<String, String> = HashMap()
        if (
            config.has(TeradataConstants.LOG_MECH) &&
                config.get(TeradataConstants.LOG_MECH).has(TeradataConstants.AUTH_TYPE) &&
                config.get(TeradataConstants.LOG_MECH).get(TeradataConstants.AUTH_TYPE).asText() !=
                    TeradataConstants.TD2_LOG_MECH
        ) {
            logmechParams[TeradataConstants.LOG_MECH] =
                config.get(TeradataConstants.LOG_MECH).get(TeradataConstants.AUTH_TYPE).asText()
        }
        return logmechParams
    }
    /**
     * Retrieves the JdbcDatabase instance based on the provided DataSource.
     *
     * @param dataSource The DataSource to create the JdbcDatabase from.
     * @return The JdbcDatabase instance.
     */
    public override fun getDatabase(dataSource: DataSource): JdbcDatabase {
        return DefaultJdbcDatabase(dataSource)
    }
    /**
     * Sets the Teradata session query band to identify the source of SQL requests originating from
     * Airbyte.
     *
     * @param jdbcDatabase The JdbcDatabase instance for which to set the query band.
     */
    private fun setQueryBand(jdbcDatabase: JdbcDatabase) {
        val setQueryBandSql =
            TeradataConstants.QUERY_BAND_SET +
                Companion.queryBand +
                TeradataConstants.QUERY_BAND_SESSION
        jdbcDatabase.execute(setQueryBandSql)
    }

    /**
     * Retrieves the default connection properties for the Teradata database based on the provided
     * configuration.
     *
     * @param config The configuration settings as a JsonNode.
     * @return A map of default connection properties.
     */
    public override fun getDefaultConnectionProperties(config: JsonNode): Map<String, String> {
        val additionalParameters: MutableMap<String, String> = HashMap()
        if (
            config.has(TeradataConstants.PARAM_SSL) &&
                config[TeradataConstants.PARAM_SSL].asBoolean()
        ) {
            if (config.has(TeradataConstants.PARAM_SSL_MODE)) {
                additionalParameters.putAll(
                    obtainConnectionOptions(config[TeradataConstants.PARAM_SSL_MODE])
                )
            } else {
                additionalParameters[TeradataConstants.PARAM_SSLMODE] = TeradataConstants.REQUIRE
            }
        }
        if (config.has(TeradataConstants.QUERY_BAND_KEY)) {
            Companion.queryBand =
                handleUserQueryBandText(
                    config[TeradataConstants.QUERY_BAND_KEY].asText(),
                )
        }
        additionalParameters[TeradataConstants.ENCRYPTDATA] = TeradataConstants.ENCRYPTDATA_ON
        return additionalParameters
    }
    /**
     * Returns a migrator that handles the migration between V1 and V2 versions of the destination.
     * This method is used to obtain an instance of a `DestinationV1V2Migrator` for migrating
     * between versions.
     *
     * @param database The database instance that is used for the migration.
     * @param databaseName The name of the database.
     * @return A `DestinationV1V2Migrator` instance specific to Teradata.
     */
    override fun getV1V2Migrator(
        database: JdbcDatabase,
        databaseName: String
    ): DestinationV1V2Migrator = TeradataV1V2Migrator(database)
    /**
     * Returns the database name extracted from the provided configuration. The database name is
     * derived from the `JdbcUtils.SCHEMA_KEY` key in the JSON configuration.
     *
     * @param config The JSON configuration containing database-related information.
     * @return A string representing the name of the database.
     */
    override fun getDatabaseName(config: JsonNode): String {
        return config[JdbcUtils.SCHEMA_KEY].asText()
    }
    /**
     * Returns a SQL generator that is used to generate SQL statements for interacting with the
     * database. This method provides a Teradata-specific SQL generator.
     *
     * @param config The JSON configuration to guide the SQL generation.
     * @return A `JdbcSqlGenerator` instance that generates SQL statements specific to Teradata.
     */
    override fun getSqlGenerator(config: JsonNode): JdbcSqlGenerator {
        return TeradataSqlGenerator()
    }
    /**
     * Returns the SQL operations for handling various SQL-related tasks, such as executing queries
     * or commands. This method returns the Teradata-specific SQL operations implementation.
     *
     * @param config The JSON configuration to guide the SQL operations.
     * @return A `SqlOperations` instance that provides SQL operations specific to Teradata.
     */
    override fun getSqlOperations(config: JsonNode): SqlOperations {
        return TeradataSqlOperations()
    }
    /**
     * Returns a generation handler that is responsible for generating database-related operations
     * such as schema generation, DDL statements, etc. This method provides a Teradata-specific
     * generation handler.
     *
     * @return A `JdbcGenerationHandler` instance that handles database generation tasks specific to
     * Teradata.
     */
    override fun getGenerationHandler(): JdbcGenerationHandler {
        return TeradataGenerationHandler()
    }
    /**
     * Returns a handler that is used to manage the destination during migration and data
     * processing. This handler is specific to Teradata and will help with raw table schema
     * handling, among other tasks.
     *
     * @param config The JSON configuration containing database and schema information.
     * @param databaseName The name of the database.
     * @param database The database instance used to interact with the destination.
     * @param rawTableSchema The raw table schema in the destination database.
     * @return A `JdbcDestinationHandler` that manages the destination database with specific
     * configurations.
     */
    override fun getDestinationHandler(
        config: JsonNode,
        databaseName: String,
        database: JdbcDatabase,
        rawTableSchema: String
    ): JdbcDestinationHandler<MinimumDestinationState> {
        return TeradataDestinationHandler(database, rawTableSchema, getGenerationHandler())
    }
    /**
     * Returns a list of migration objects that perform database migrations. In this case, an empty
     * list is returned, indicating that no migrations are needed for this particular database.
     *
     * @param database The database instance being migrated.
     * @param databaseName The name of the database being migrated.
     * @param sqlGenerator The SQL generator used for creating SQL statements during migrations.
     * @param destinationHandler The handler used to manage the destination during migration.
     * @return An empty list, as no migrations are specified.
     */
    override fun getMigrations(
        database: JdbcDatabase,
        databaseName: String,
        sqlGenerator: SqlGenerator,
        destinationHandler: DestinationHandler<MinimumDestinationState>
    ): List<Migration<MinimumDestinationState>> {
        return emptyList()
    }
    /**
     * Indicates whether the destination is considered a V2 destination. This flag can be used to
     * determine whether the system is operating in a V1 or V2 context.
     *
     * @return True if the destination is V2, false otherwise.
     */
    override val isV2Destination: Boolean
        get() = true

    /**
     * Obtains additional connection options like SSL configuration.
     *
     * @param encryption The JsonNode containing SSL parameters.
     * @return A map of additional connection properties.
     */
    private fun obtainConnectionOptions(encryption: JsonNode): Map<String, String> {
        val additionalParameters: MutableMap<String, String> = HashMap()
        if (!encryption.isNull) {
            val method = encryption[TeradataConstants.PARAM_MODE].asText()
            additionalParameters[TeradataConstants.PARAM_SSLMODE] = method

            if (TeradataConstants.VERIFY_CA == method || TeradataConstants.VERIFY_FULL == method) {
                createCertificateFile(encryption[TeradataConstants.CA_CERT_KEY].asText())
                additionalParameters[TeradataConstants.PARAM_SSLCA] =
                    TeradataConstants.CA_CERTIFICATE
            }
        }
        return additionalParameters
    }

    /** Creates certificate file for verify-ca and verify-full ssl connection */
    @Throws(IOException::class)
    private fun createCertificateFile(fileValue: String) {
        PrintWriter(TeradataConstants.CA_CERTIFICATE, StandardCharsets.UTF_8).use { out ->
            out.print(fileValue)
        }
    }
    /**
     * Handles and validates the user-defined query band text.
     *
     * @param queryBandText The user-defined query band text.
     * @return The validated query band text, ensuring required parameters are presentin required
     * format.
     */
    private fun handleUserQueryBandText(queryBandText: String?): String {
        if (queryBandText.isNullOrBlank()) {
            return Companion.queryBand
        }
        var updatedQueryBand = StringBuilder(queryBandText)
        // checking org doesn't exist in query_band, appending 'org=teradata-internal-telem'
        // If it exists, user might have set some value of their own, so doing nothing in that case
        val orgMatcher = Pattern.compile("org\\s*=").matcher(queryBandText)
        if (!orgMatcher.find()) {
            if (queryBandText.isNotBlank() && !queryBandText.endsWith(";")) {
                updatedQueryBand.append(";")
            }
            updatedQueryBand.append(TeradataConstants.DEFAULT_QUERY_BAND_ORG)
        }

        // Ensure appname contains airbyte is present or append it if it exists with different value
        val appNameMatcher = Pattern.compile("appname\\s*=\\s*([^;]*)").matcher(updatedQueryBand)
        if (appNameMatcher.find()) {
            val appNameValue = appNameMatcher.group(1).trim { it <= ' ' }
            if (!appNameValue.lowercase(Locale.getDefault()).contains("airbyte")) {
                updatedQueryBand =
                    StringBuilder(
                        updatedQueryBand
                            .toString()
                            .replace(
                                "appname\\s*=\\s*([^;]*)".toRegex(),
                                "appname=" + appNameValue + "_airbyte",
                            ),
                    )
            }
        } else {
            if (updatedQueryBand.isNotEmpty() && !updatedQueryBand.toString().endsWith(";")) {
                updatedQueryBand.append(";")
            }
            updatedQueryBand.append(TeradataConstants.DEFAULT_QUERY_BAND_APPNAME)
        }
        return updatedQueryBand.toString()
    }
    /**
     * Converts the provided configuration into JDBC configuration settings.
     *
     * @param config The configuration settings as a JsonNode.
     * @return The converted JsonNode containing JDBC configuration.
     */
    override fun toJdbcConfig(config: JsonNode): JsonNode {
        val schema = config[JdbcUtils.SCHEMA_KEY]?.asText() ?: TeradataConstants.DEFAULT_SCHEMA_NAME
        val jdbcUrl = String.format("jdbc:teradata://%s/", config[JdbcUtils.HOST_KEY].asText())
        val userName = config[TeradataConstants.LOG_MECH]?.get(JdbcUtils.USERNAME_KEY)?.asText()
        val password = config[TeradataConstants.LOG_MECH]?.get(JdbcUtils.PASSWORD_KEY)?.asText()
        val configBuilder =
            ImmutableMap.builder<Any, Any>()
                .put(JdbcUtils.JDBC_URL_KEY, jdbcUrl)
                .put(JdbcUtils.SCHEMA_KEY, schema)
        userName?.let { configBuilder.put(JdbcUtils.USERNAME_KEY, it) }
        password?.let { configBuilder.put(JdbcUtils.PASSWORD_KEY, it) }
        config[JdbcUtils.JDBC_URL_PARAMS_KEY]?.asText()?.let {
            configBuilder.put(JdbcUtils.JDBC_URL_PARAMS_KEY, it)
        }
        return Jsons.jsonNode(configBuilder.build())
    }

    val queryBand: String
        get() = Companion.queryBand

    companion object {

        private var queryBand = TeradataConstants.DEFAULT_QUERY_BAND

        @Throws(Exception::class)
        @JvmStatic
        fun main(args: Array<String>) {
            IntegrationRunner(TeradataDestination()).run(args)
        }
    }
}
